File: internal/protocol/protocol.go
// Package protocol defines the wire protocol for WebSocket binary
// frames. Mirrors the TypeScript implementation in
// packages/kagami/src/protocol.ts. Frame format:
// [4-byte header length][JSON header][raw body bytes].
// Handles serialization, deserialization, and chunking for bodies
// exceeding the 1 MiB frame limit.
package protocol

import (
	"encoding/binary"
	"encoding/json"
	"errors"
	"fmt"
)

// ChunkSize is the default maximum body size per frame (512 KB).
const ChunkSize = 524288

// Known message types for validation.
var knownTypes = map[string]bool{
	"http_request":    true,
	"http_response":   true,
	"http_body_chunk": true,
	"ping":            true,
	"pong":            true,
	"error":           true,
}

// EncodeFrame serializes a JSON-marshalable header and optional body
// into a binary wire frame: [4-byte header length (uint32 BE)][JSON header][body].
func EncodeFrame(header any, body []byte) ([]byte, error) {
	headerJSON, err := json.Marshal(header)
	if err != nil {
		return nil, fmt.Errorf("marshal header: %w", err)
	}

	headerLen := len(headerJSON)
	frame := make([]byte, 4+headerLen+len(body))
	binary.BigEndian.PutUint32(frame[:4], uint32(headerLen))
	copy(frame[4:4+headerLen], headerJSON)
	copy(frame[4+headerLen:], body)
	return frame, nil
}

// DecodeFrame splits a binary wire frame into its JSON header bytes and
// raw body bytes. Returns an error if the frame is too short or the
// header length exceeds the frame.
func DecodeFrame(data []byte) (Frame, error) {
	if len(data) < 4 {
		return Frame{}, errors.New("frame too short: missing header length")
	}

	headerLen := binary.BigEndian.Uint32(data[:4])
	if int(headerLen) > len(data)-4 {
		return Frame{}, fmt.Errorf("header length %d exceeds remaining frame size %d", headerLen, len(data)-4)
	}

	header := data[4 : 4+headerLen]
	var body []byte
	if remaining := data[4+headerLen:]; len(remaining) > 0 {
		body = remaining
	}

	return Frame{Header: header, Body: body}, nil
}

// ParseHeaderType extracts the "type" field from raw JSON header bytes
// and validates it against known message types.
func ParseHeaderType(header []byte) (string, error) {
	var envelope struct {
		Type string `json:"type"`
	}
	if err := json.Unmarshal(header, &envelope); err != nil {
		return "", fmt.Errorf("unmarshal header type: %w", err)
	}
	if envelope.Type == "" {
		return "", errors.New("missing type field in header")
	}
	if !knownTypes[envelope.Type] {
		return "", fmt.Errorf("unknown message type: %q", envelope.Type)
	}
	return envelope.Type, nil
}

// ChunkBody splits a large body into multiple encoded wire frames ready
// for WebSocket transmission. The first frame carries the original header
// (with Chunked set to true) and the first body segment. Subsequent
// frames use HttpBodyChunkHeader with the same message ID. The last
// chunk has Final set to true.
//
// If the body fits within chunkSize, a single encoded frame is returned
// with the header unmodified (Chunked remains false/omitted).
//
// The header must be an *HttpRequestHeader or *HttpResponseHeader.
// The caller's header struct is not mutated.
func ChunkBody(header any, body []byte, chunkSize int) ([][]byte, error) {
	if chunkSize <= 0 {
		return nil, errors.New("chunk size must be positive")
	}

	if len(body) <= chunkSize {
		encoded, err := EncodeFrame(header, body)
		if err != nil {
			return nil, err
		}
		return [][]byte{encoded}, nil
	}

	// Copy the header and set Chunked flag, extract the message ID.
	var (
		chunkedHeader any
		msgID         string
	)
	switch h := header.(type) {
	case *HttpRequestHeader:
		cp := *h
		cp.Chunked = true
		chunkedHeader = &cp
		msgID = h.ID
	case *HttpResponseHeader:
		cp := *h
		cp.Chunked = true
		chunkedHeader = &cp
		msgID = h.ID
	default:
		return nil, fmt.Errorf("unsupported header type for chunking: %T", header)
	}

	var frames [][]byte

	// First frame: original header (with chunked) + first chunk.
	firstChunk := body[:chunkSize]
	encoded, err := EncodeFrame(chunkedHeader, firstChunk)
	if err != nil {
		return nil, err
	}
	frames = append(frames, encoded)

	// Continuation chunks.
	remaining := body[chunkSize:]
	for len(remaining) > 0 {
		end := min(chunkSize, len(remaining))
		chunk := remaining[:end]
		remaining = remaining[end:]

		chunkHeader := &HttpBodyChunkHeader{
			MessageHeader: MessageHeader{
				Type: "http_body_chunk",
				ID:   msgID,
			},
			Final: len(remaining) == 0,
		}

		encoded, err := EncodeFrame(chunkHeader, chunk)
		if err != nil {
			return nil, err
		}
		frames = append(frames, encoded)
	}

	return frames, nil
}

// ReassembleChunks concatenates body byte slices into the original
// complete body. The receiver decodes each wire frame and passes
// the body portions directly.
func ReassembleChunks(chunks [][]byte) []byte {
	size := 0
	for _, c := range chunks {
		size += len(c)
	}
	result := make([]byte, 0, size)
	for _, c := range chunks {
		result = append(result, c...)
	}
	return result
}

File: packages/kagami/src/protocol.ts
/**
 * Wire protocol types and serialization for kagami tunnels.
 *
 * All messages are sent as WebSocket binary frames:
 *   [4-byte header length (uint32 BE)][JSON header][raw body bytes]
 *
 * Control messages (ping, pong, error) have no body.
 * Large bodies are chunked across multiple frames.
 */

// --- Message Header Types ---

export interface MessageHeader {
  type:
    | "http_request"
    | "http_response"
    | "http_body_chunk"
    | "ping"
    | "pong"
    | "error";
  id: string;
}

export interface HttpRequestHeader extends MessageHeader {
  type: "http_request";
  method: string;
  host: string;
  path: string;
  query: string;
  headers: Record<string, string[]>;
  chunked?: boolean;
}

export interface HttpResponseHeader extends MessageHeader {
  type: "http_response";
  status: number;
  headers: Record<string, string[]>;
  chunked?: boolean;
}

export interface HttpBodyChunkHeader extends MessageHeader {
  type: "http_body_chunk";
  final: boolean;
}

export interface PingHeader extends MessageHeader {
  type: "ping";
}

export interface PongHeader extends MessageHeader {
  type: "pong";
}

export interface ErrorHeader extends MessageHeader {
  type: "error";
  code: string;
  message: string;
}

export type TunnelHeader =
  | HttpRequestHeader
  | HttpResponseHeader
  | HttpBodyChunkHeader
  | PingHeader
  | PongHeader
  | ErrorHeader;

// --- Constants ---

/** Default chunk size for splitting large bodies (512 KB). */
export const DEFAULT_CHUNK_SIZE = 524288;

// --- Serialization / Deserialization ---

/**
 * Encode a header + optional body into a binary wire frame.
 *
 * Frame format:
 *   [4-byte header length (uint32 BE)][JSON header (UTF-8)][body (raw bytes)]
 */
export function encodeFrame(
  header: TunnelHeader,
  body?: Uint8Array,
): ArrayBuffer {
  const encoder = new TextEncoder();
  const headerBytes = encoder.encode(JSON.stringify(header));
  const headerLen = headerBytes.byteLength;
  const bodyLen = body ? body.byteLength : 0;

  const frame = new ArrayBuffer(4 + headerLen + bodyLen);
  const view = new DataView(frame);

  // 4-byte header length, big-endian
  view.setUint32(0, headerLen, false);

  // JSON header bytes
  const frameBytes = new Uint8Array(frame);
  frameBytes.set(headerBytes, 4);

  // Body bytes (if any)
  if (body && bodyLen > 0) {
    frameBytes.set(body, 4 + headerLen);
  }

  return frame;
}

/**
 * Decode a binary wire frame into header + body.
 *
 * Throws if the frame is malformed (too short, bad header length,
 * missing required fields).
 */
export function decodeFrame(data: ArrayBuffer): {
  header: TunnelHeader;
  body: Uint8Array;
} {
  if (data.byteLength < 4) {
    throw new Error("Frame too short: missing header length");
  }

  const view = new DataView(data);
  const headerLen = view.getUint32(0, false);

  if (4 + headerLen > data.byteLength) {
    throw new Error(
      `Header length ${headerLen} exceeds frame size ${data.byteLength}`,
    );
  }

  const decoder = new TextDecoder();
  const headerBytes = new Uint8Array(data, 4, headerLen);
  const headerJson = decoder.decode(headerBytes);

  let parsed: unknown;
  try {
    parsed = JSON.parse(headerJson);
  } catch {
    throw new Error("Invalid JSON in frame header");
  }

  if (typeof parsed !== "object" || parsed === null) {
    throw new Error("Frame header is not a JSON object");
  }

  const obj = parsed as Record<string, unknown>;

  if (typeof obj.type !== "string") {
    throw new Error("Missing or invalid 'type' field in frame header");
  }

  if (typeof obj.id !== "string") {
    throw new Error("Missing or invalid 'id' field in frame header");
  }

  const header = obj as unknown as TunnelHeader;
  const body = new Uint8Array(data, 4 + headerLen);

  return { header, body };
}

// --- Chunking ---

/** Result of chunking a message: the initial frame plus any continuation frames. */
export interface ChunkedFrames {
  /** All frames in order. First is the initial http_request/http_response, rest are http_body_chunk. */
  frames: ArrayBuffer[];
}

/**
 * Encode a header + body, splitting into chunked frames if the body
 * exceeds chunkSize.
 *
 * For small bodies (<= chunkSize), returns a single frame with no
 * chunked flag. For large bodies, returns:
 *   1. Initial frame with `chunked: true` and first chunk as body
 *   2. N-1 continuation `http_body_chunk` frames
 *   3. Final continuation frame with `final: true`
 */
export function encodeChunked(
  header: HttpRequestHeader | HttpResponseHeader,
  body: Uint8Array,
  chunkSize: number = DEFAULT_CHUNK_SIZE,
): ChunkedFrames {
  if (body.byteLength <= chunkSize) {
    // Single frame, no chunked flag
    return { frames: [encodeFrame(header, body)] };
  }

  const frames: ArrayBuffer[] = [];
  let offset = 0;

  // First frame: initial header with chunked: true + first chunk
  const firstChunk = body.slice(offset, offset + chunkSize);
  offset += chunkSize;
  const chunkedHeader = { ...header, chunked: true as const };
  frames.push(encodeFrame(chunkedHeader, firstChunk));

  // Continuation frames
  while (offset < body.byteLength) {
    const end = Math.min(offset + chunkSize, body.byteLength);
    const chunk = body.slice(offset, end);
    const isFinal = end >= body.byteLength;

    const chunkHeader: HttpBodyChunkHeader = {
      type: "http_body_chunk",
      id: header.id,
      final: isFinal,
    };

    frames.push(encodeFrame(chunkHeader, chunk));
    offset = end;
  }

  return { frames };
}

/**
 * Reassemble chunked body data from multiple frames.
 *
 * Takes an array of body Uint8Arrays (in order: first chunk from initial
 * frame, then each continuation chunk) and concatenates them.
 */
export function reassembleChunks(chunks: Uint8Array[]): Uint8Array {
  if (chunks.length === 0) {
    return new Uint8Array(0);
  }

  if (chunks.length === 1) {
    return chunks[0];
  }

  let totalLen = 0;
  for (const chunk of chunks) {
    totalLen += chunk.byteLength;
  }

  const result = new Uint8Array(totalLen);
  let offset = 0;
  for (const chunk of chunks) {
    result.set(chunk, offset);
    offset += chunk.byteLength;
  }

  return result;
}
